pygbm/loss.py

```diff
@@ -0,0 +1,97 @@
+from abc import ABC
+
+import numpy as np
+from numba import njit, prange
+import numba
+
+
+@njit
+def _get_threads_chunks(total_size):
+    # Divide [0, total_size - 1] into n_threads contiguous regions, and
+    # returns the starts and ends of each region. Used to simulate a 'static'
+    # scheduling.
+    n_threads = numba.config.NUMBA_DEFAULT_NUM_THREADS
+    sizes = np.full(n_threads, total_size // n_threads, dtype=np.int32)
+    sizes[:total_size % n_threads] += 1
+    starts = np.zeros(n_threads, dtype=np.int32)
+    starts[1:] = np.cumsum(sizes[:-1])
+    ends = starts + sizes
+
+    return starts, ends, n_threads
```

Comment:
It would be great to contact the numba core devs to get their opinion on this kind of threading-scheduling boiler plate code.

Also, shouldn't we use `n_chunks = 3 * n_threads` to mitigate potential stragglers?

Reply:
In particular it would be great if numba could provided some kind of ["guided" scheduling strategy](http://jakascorner.com/blog/2016/06/omp-for-scheduling.html#guided) like openmp does, in particular with the ability to pass a minimum chunk size.

Reply:
I commented https://github.com/numba/numba/issues/2649#issuecomment-442200409 about this a few days ago